lms/services/assignment.py

```diff
@@ -10,55 +13,149 @@ def __init__(self, db):
-            .filter_by(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                resource_link_id=resource_link_id,
-            )
-            .one_or_none()
+    def get(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        """Get an assignment using using resource_link_id, ext_lti_assignment_id or both."""
+
+        query = self._db.query(Assignment).filter(
+            Assignment.tool_consumer_instance_guid == tool_consumer_instance_guid
         )
+        if resource_link_id and not ext_lti_assignment_id:
+            # Non canvas assignments
+            query = query.filter(Assignment.resource_link_id == resource_link_id)
+        elif not resource_link_id and ext_lti_assignment_id:
+            # creating/editing a canvas assignment
+            query = query.filter(
+                Assignment.ext_lti_assignment_id == ext_lti_assignment_id
+            )
+        elif resource_link_id and ext_lti_assignment_id:
+            query = query.filter(
+                (
+                    # Regular canvas launch
+                    (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                    # Configuring a file assignment that was stored in the DB
+                    # before all canvas assignments were stored in the DB
+                    | (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id.is_(None))
+                    )
+                    #  First launch of a newly configured assignment
+                    | (
+                        (Assignment.resource_link_id.is_(None))
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                )
+            )
+        else:
+            log.exception(
+                "Can't get an assignment without neither resource_link_id or ext_lti_assignment_id"
+            )
+            return None
+
+        assignments = query.all()
+        if not assignments:
+            return None
+
+        if len(assignments) == 2:
+            # We stored a canvas file assignment before (storing its resource_link_id)
+            # we later configured it (storing its ext_lti_assignment_id) in a new row
+            # and now we are launching it, we want to merge those two assignments
+            old_assignment = (
+                assignments[0] if assignments[0].resource_link_id else assignments[1]
+            )
+            assignment = (
+                assignments[0]
+                if assignments[0].ext_lti_assignment_id
+                else assignments[1]
+            )
+            self._db.delete(old_assignment)
+            self._db.flush()
```

Comment:
We need this because there are already canvas file assignments in the DB which have a `resource_link_id`.

After this, re-configured assignments will be created with `ext_lti_assignment_id` creating a duplicate.